src/service.rs

```diff
@@ -713,5 +764,173 @@ mod test {
+                            assert_eq!(cir.result_token, PREPARE_CI_TOKEN);
+                            unwrap_result!(cir.result)
+                        }
+                        m => panic!("Received unexpected event: m == {:?}", m),
+                    };
+                    let their_ci = our_ci.to_their_connection_info();
+                    ci_tx.send(their_ci);
+                    self.our_cis.push(our_ci);
+                }
+            }
+
+            fn run(self, barrier: Arc<Barrier>) -> JoinHandle<()> {
+                thread!("run!", move || {
+                    for (our_ci, their_ci) in self.our_cis.into_iter().zip(self.connection_id_rx.into_iter()) {
+                        self.service.connect(our_ci, their_ci);
+                    }
+                    let mut their_ids = HashMap::new();
+                    for _ in 0..(NUM_SERVICES - 1) {
+                        let their_id = match unwrap_result!(self.event_rx.recv()) {
+                            Event::NewPeer(Ok(()), their_id) => their_id,
+                            m => panic!("Expected NewPeer message. Got message {:?}", m),
+                        };
+                        match their_ids.insert(their_id, 0u32) {
+                            Some(_) => panic!("Recieved two NewPeer events for same peer!"),
+                            None => (),
+                        };
+                    }
+
+                    for their_id in their_ids.keys() {
+                        for n in 0..NUM_MSGS {
+                            let mut msg = Vec::with_capacity(MSG_SIZE);
+                            for _ in 0..MSG_SIZE {
+                                msg.push(n as u8);
+                            }
+                            self.service.send(their_id, msg);
+                        };
+                    }
+
+                    for _ in 0..((NUM_SERVICES - 1) * NUM_MSGS) {
+                        match unwrap_result!(self.event_rx.recv()) {
+                            Event::NewMessage(their_id, msg) => {
+                                let n = msg[0];
+                                assert_eq!(msg.len(), MSG_SIZE);
+                                for m in msg {
+                                    assert_eq!(n, m);
+                                };
+                                match their_ids.entry(their_id.clone()) {
+                                    hash_map::Entry::Occupied(mut oe) => {
+                                        let next_msg = oe.get_mut();
+                                        assert_eq!(*next_msg, n as u32);
+                                        *next_msg += 1;
+                                    },
+                                    hash_map::Entry::Vacant(ve) => panic!("impossible!"),
+                                }
+                            },
+                            m => panic!("ERROR ERROR {} Unexpected msg receiving NewMessage: {:?}", ::time::SteadyTime::now(), m),
```

Comment:
With `use time::SteadyTime` and without the "ERROR ERROR" this would probably properly fit in the line.